keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,1161 @@
+kind: ClusterNetworkPolicy
+spec:
+  priority: 100
+  - appliedTo:
+     namespaceSelector: {}
+    ingress:
+    - action: Allow
+      from:
+      - namespaces:
+          scope: self
+        workloadSelector:
+          podSelector:
+            matchLabels:
+              app: b
+```
+
+The above ClusterNetworkPolicy should be interpreted as: for each Namespace in
+the cluster, all Pods in that Namespace should strictly allow traffic from Pods in
+the _same Namespace_ who has label app=b. Hence, the policy above allows
+x/b1 -> x/a1 and y/b2 -> y/a2, but does not allow y/b2 -> x/a1 and x/b1 -> y/a2.
+
+__SameLabels:__
+This is a special strategy to indicate that the rule only applies to the Namespaces
+which share the same label value. Since the Pods selected by the ClusterNetworkPolicy `appliedTo`
+could be from multiple Namespaces, the scope of ingress/egress rules whose `scope=samelabels; labels: [tenant]`
+will be all the Pods from the Namespaces who have the same label value for the "tenant" key.
+Consider the following example:
+
+- Pods [a1, b1] exist in Namespace t1-ns1, which has label `tenant=t1`.
+- Pods [a2, b2] exist in Namespace t1-ns2, which has label `tenant=t1`.
+- Pods [a3, b3] exist in Namespace t2-ns1, which has label `tenant=t2`.
+- Pods [a4, b4] exist in Namespace t2-ns2, which has label `tenant=t2`.
+
+```yaml
+apiVersion: netpol.networking.k8s.io/v1alpha1
+kind: ClusterNetworkPolicy
+spec:
+  priority: 50
+  - appliedTo:
+     namespaceSelector:
+       matchExpressions: {key: "tenant"; operator: Exists}
+    ingress:
+     - action: Pass
+       from:
+       - namespaces:
+          scope: samelabels
+          labels:
+            - tenant
+```
+
+The above ClusterNetworkPolicy should be interpreted as: for each Namespace in
+the cluster who has a label key set as "tenant", traffic for all Pods in that Namespace
+from all Pods in the Namespaces who has the same label value for key `tenant` is delegated to the Namespace
+admins, i.e such traffic will not be subject to any CNP (`priority` > 50) rules and be evaluated by K8s NetworkPolicies.
+Hence, the policy above delegates traffic from all Pods in Namespaces labeled `tenant=t1` i.e. t1-ns1 and t1-ns2,
+to reach each other, to K8s NetworkPolicies, similarly traffic for all Pods in Namespaces labeled `tenant=t2`
+i.e. t2-ns1 and t2-ns2, to talk to each other is delegated to K8s NetworkPolicies as well, however it does not
+delegate traffic from any Pod in t1-ns1 or t1-ns2 to reach Pods in t2-ns1 or t2-ns2, such traffic is still subject
+to CNP rules.
+
+### Sample Specs for User Stories
+
+#### Sample spec for Story 1: Deny traffic at a cluster level
+
+```yaml
+apiVersion: netpol.networking.k8s.io/v1alpha1
+kind: ClusterNetworkPolicy
+spec:
+  priority: 10
+  - appliedTo:
+     namespaceSelector: {}
+    ingress:
+     - action: Deny
+       from:
+       - namespaceSelector:
+          matchLabels:
+            isolatedByAdmin: true
+```
```

Comment:
So this says "for every pod in every namespace, disallow ingress from `isolatedByAdmin: true` namespaces". It would most likely be more efficient to flip it around and say "for every pod in every `isolatedByAdmin: true` namespace, denying all egress":

```
apiVersion: netpol.networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
spec:
  priority: 10
  - appliedTo:
      namespaceSelector:
        matchLabels:
          isolatedByAdmin: true
    egress:
     - action: Deny
```